aya/src/programs/probe.rs

```diff
@@ -23,10 +29,76 @@ pub enum ProbeKind {
 pub(crate) fn attach(
     program_data: &mut ProgramData,
     kind: ProbeKind,
-    name: &str,
+    fn_name: &str,
     offset: u64,
     pid: Option<pid_t>,
 ) -> Result<LinkRef, ProgramError> {
+    // https://github.com/torvalds/linux/commit/e12f03d7031a977356e3d7b75a68c2185ff8d155
+    // Use debugfs to create probe
+    let k_ver = kernel_version().unwrap();
+    if k_ver < (4, 17, 0) {
+        let (fd, event_alias) = create_as_trace_point(kind, fn_name, offset, pid)?;
+
+        return perf_attach_debugfs(program_data, fd, kind, event_alias);
+    };
+
+    let fd = create_as_probe(kind, fn_name, offset, pid)?;
+
+    perf_attach(program_data, fd)
+}
+
+pub(crate) fn detach_debug_fs(kind: ProbeKind, event_alias: &str) -> Result<(), ProgramError> {
+    use ProbeKind::*;
+
+    /*
+     * Taken from https://github.com/iovisor/bcc/blob/67f59ee80fcf5deedaacba1436d9fa09d32a16a0/src/cc/libbpf.c#L1173
+     *
+     * For [k,u]probe created with perf_event_open (on newer kernel), it is
+     * not necessary to clean it up in [k,u]probe_events. We first look up
+     * the %s_bcc_%d line in [k,u]probe_events. If the event is not found,
+     * it is safe to skip the cleaning up process (write -:... to the file).
+     */
+
+    let event_type = match kind {
+        KProbe | KRetProbe => "kprobe",
+        UProbe | URetProbe => "uprobe",
+    };
+
+    let events_file_name = format!("/sys/kernel/debug/tracing/{}_events", event_type);
+
```

Comment:
I would open the file here, then make `find_in_sys_kernel_debug_tracing_events` and `delete_in_sys_kernel_debug_tracing_events` closures

Reply:
The behaviour of libbpf is to read the file with `O_RDONLY` and then close and reopen the file with `O_APPEND`. I'm not sure if that's required by the kernel interface, but that's what I was trying to emulate.

Reply:
I would be very surprised if that was the case!